LoRaEngine/modules/LoRaWanNetworkSrvModule/Logger/LogSink.cs

```diff
@@ -0,0 +1,295 @@
+        public static ILogSink Create(ILogSink first, ILogSink second) =>
+            new LogSink(first ?? throw new ArgumentNullException(nameof(first)),
+                        second ?? throw new ArgumentNullException(nameof(second)));
+
+        public static ILogSink? Choose(ILogSink? first, ILogSink? second) =>
+            (first, second) switch
+            {
+                ({ } fst, null) => fst,
+                (null, { } snd) => snd,
+                ({ } fst, { } snd) => Create(fst, snd),
+                _ => null
+            };
+
+        private sealed class LogSink : ILogSink, IDisposable
+        {
+            private readonly ILogSink first;
+            private readonly ILogSink second;
+
+            public LogSink(ILogSink first, ILogSink second)
+            {
+                LogLevel = (LogLevel)Math.Min((int)first.LogLevel, (int)second.LogLevel);
+                (this.first, this.second) = (first, second);
+            }
+
+            public LogLevel LogLevel { get; }
+
+            public void Log(LogLevel logLevel, string message)
+            {
+                this.first.Log(logLevel, message);
+                this.second.Log(logLevel, message);
+            }
+
+            public void Dispose()
+            {
+                (this.first as IDisposable)?.Dispose();
+                (this.second as IDisposable)?.Dispose();
+            }
+        }
+    }
+
+    public sealed class ConsoleLogSink : LogSink
+    {
+        public ConsoleLogSink(LogLevel logLevel) : base(logLevel) { }
+
+        protected override void CoreLog(LogLevel logLevel, string message)
+        {
+            var writer = logLevel == LogLevel.Error ? Console.Error : Console.Out;
+            writer.WriteLine(FormattableString.Invariant($"{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff} {message}"));
+        }
+    }
+
+    public sealed class TcpLogSink : ILogSink, IDisposable
+    {
+        private readonly ILogger? logger;
+        private readonly IPEndPoint serverEndpoint;
+        private readonly Func<string, string>? formatter;
+        private readonly int maxRetryAttempts;
+        private readonly TimeSpan retryDelay;
+        private readonly CancellationTokenSource cancellationTokenSource = new();
+        private readonly Channel<string> channel;
+
+        public static TcpLogSink Start(IPEndPoint serverEndPoint, LogLevel logLevel,
+                                       int? maxRetryAttempts = null,
+                                       TimeSpan? retryDelay = null,
+                                       int? backlogCapacity = null,
+                                       Func<string, string>? formatter = null,
+                                       ILogger<TcpLogSink>? logger = null)
+        {
+            var sink = new TcpLogSink(serverEndPoint, logLevel, maxRetryAttempts, retryDelay, backlogCapacity, formatter, logger);
+            _ = Task.Run(() => sink.SendAllLogMessagesAsync(sink.cancellationTokenSource.Token));
```

Comment:
What's the reason why sending log messages should happen only on a best effort basis? How do we observe/debug if something goes wrong with sending logs on the `TcpLogSink`?

Reply:
`SendAllLogMessagesAsync` logs emits debug, warning and error messages when things go wrong.